extensions/amp-video/0.1/amp-video.js

```diff
@@ -204,38 +243,131 @@ class AmpVideo extends AMP.BaseElement {
       return Promise.resolve();
     }
 
-    if (this.element.getAttribute('src')) {
-      assertHttpsUrl(this.element.getAttribute('src'), this.element);
-    }
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
 
-    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT, this.video_,
+    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT,
+        dev().assertElement(this.video_),
         /* opt_removeMissingAttrs */ true);
 
-    this.getRealChildNodes().forEach(child => {
-      // Skip the video we already added to the element.
-      if (this.video_ === child) {
-        return;
-      }
-      if (child.getAttribute && child.getAttribute('src')) {
-        assertHttpsUrl(child.getAttribute('src'),
-            dev().assertElement(child));
+    if (viewer.getVisibilityState() == VisibilityState.PRERENDER) {
+      if (!this.element.hasAttribute('preload')) {
+        this.video_.setAttribute('preload', 'auto');
       }
-      this.video_.appendChild(child);
-    });
+      this.propagatePrerenderSources_();
```

Comment:
Nit: seems like you could move this this.propagatePrerenderSources_() before the if/else?

Reply:
done

Reply:
done

Reply:
also renamed to `propagateCachedSources_`, seems less confusing and added some comments.